### PR TITLE
Use Jinja2 to parse layout files (editable packages)

### LINKED ISSUE #4424: Use jinja2 templates for editable layouts
Syntax in editable layouts are currently using placeholders like ``{settings.build_type}`` to make some conditional logic regarding folder paths.

We knew it has some limitations specially with placeholder like ``{options.shared}``, that will replace the value with a ``True``/``False`` instead of a useful value.

We evaluated the following:
- Inline evaluation of python code
- Evaluation of field headers like ``[includedirs#settings.build_type==True]``
- Python syntax in *layout.py* implementing a ``package_info()``
- Jinja2 templating

And decided that jinja2 is a fairly common orthogonal language used in a lot of tools, syntax is well documented and contemplates many conditional scenarios. Also this should be quite easy to implement.

The information available in this jinja2 templates should be ``options`` and ``settings``.

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Feature: Apply Jinja2 to layout files before parsing them
Docs: https://github.com/conan-io/docs/pull/1093

This PR executes Jinja2 over layout files (context contains `reference`, `settings` and `options`) before parsing those files so the user will be able to apply some logic on the constructed path to match the underlying filesystem. As this logic depends on `settings` and `options` we are no longer able to parse these files before knowing the node/settings/options it is going to be applied to.

closes #4424 

Running on _more_ python versions as we are introducing a new dependency:
@PYVERS: Macos@py27, Windows@py36, Linux@py27, py34

### CODE REVIEW

File: conans/model/editable_cpp_info.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
 # coding=utf-8
 import os
 from collections import OrderedDict
+from io import StringIO
```
Comment (lasote) [inline conans/model/editable_cpp_info.py]: Probably better to use `six.StringIO`. Change also the import from `utils.tools.py` (not related to this PR, but please do it)

File: conans/model/editable_cpp_info.py [hunk 2]
```diff
@@ -35,68 +36,76 @@ class EditableLayout(object):
...
-        try:
-            return self._work_on_item(path, settings, options)
-        except Exception as e:
-            raise ConanException("Error getting fHolder '%s' from layout: %s" % (str(name), str(e)))
 
     @staticmethod
-    def load(filepath):
-        parser = configparser.ConfigParser(allow_no_value=True)
-        parser.optionxform = str
+    def _work_on_item(value):
```
Comment (lasote) [inline conans/model/editable_cpp_info.py]: Maybe now it deserves a new name or to disappear.

File: conans/test/functional/editable/link_create_test.py [hunk 3]
```diff
@@ -48,6 +48,19 @@ def test_editable_list_search(self):
         t.run("search")
         self.assertIn("lib/version@user/name", t.out)
 
+    def test_editable_list_search(self):
```
Comment (memsharded) [inline conans/test/functional/editable/link_create_test.py]: This test is duplicated with the above

**Comment by jgsogo** [comment]
There are just the needed changes:
 * Functionality: it all is in the class `EditableLayout`. Introducing the Jinja2 template engine requires modifying some logi: we can't parse the file in advance, we need to wait until the last moment to know settings and options (it is not a refactor, it is needed).
 * Tests:
    - functional/integration: only minor changes to adopt the new format
    - unittests: one function <-> one class::function, not so hard to follow:
        + `apply_test.py` ==> `EditableLayout::apply_to`
        + `load_data_test.py` ==> `EditableLayout::_load_data` (focusing on field errors)
        + `parse_test.py` ==> `EditableLayout::_load_data`  (focusing on render+parsing)
        + `work_on_items_test.py` ==> Removed, as there is nothing to test now, I agree with @lasote and I would remove this function

If the functionality has changed, the unittest change, there is no _diff_ suitable for them.

File: conans/requirements.txt [hunk 4]
```diff
@@ -13,3 +13,4 @@ pygments>=2.0, <3.0
 astroid>=1.6.5
 deprecation>=2.0, <2.1
 tqdm>=4.28.1, <5
+Jinja2==2.10
```
Comment (memsharded) [inline conans/requirements.txt]: Maybe this requirement could be more relaxed? So it minimizes collisions with other possible Jinja2 previously installed.

Comment (jgsogo) [reply conans/requirements.txt]: Upps, yes, didn't realize about it after adding it to the requirements.

Quick testing: `sphinx` requires `>=2.3` and our tests are working with it... and Jinja2.3 is from Feb 10, 2010. Maybe go for `Jinja2>=2.3, <3`?
